language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+    ...
+    newvalue(:mask, :event => :service_disabled, :required_features => :maskable) do
+      provider.mask
+    end
+  end
+end
+```
+
+### mixins
+
+Since puppet resource types are defined in ruby, you can mixin additional functionality. The AWS module uses this to create subclasses for different types of Route53 DNS records:
+
+```ruby
+Puppet::Type.newtype(:route53_a_record) do
+  extend PuppetX::Puppetlabs::Route53Record
+  @doc = 'Type representing a Route53 DNS record.'
+  create_properties_and_params()
+end
+```
+
+where the base module defines common parameters and properties for all `Route53Records`:
+
+```ruby
+module PuppetX
+  module Puppetlabs
+    module Route53Record
+      def create_properties_and_params
+        ensurable
+        newproperty(:zone) do
+        end
+      end
+    end
+  end
+end
+```
+
+## Parameter & Property DSL Methods
+
+The following DSL methods are commonly used for individual parameters and properties within the body of `newparam` and `newproperty` respectively:
+
+### desc
+
+Description of the property, output when running `puppet describe <type>`:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:region) do
+    desc 'The region in which to launch the instance.'
+  end
+end
+```
+
+### isnamevar
+
+Specifies that the parameter is the `namevar` (aka identity) for the resource:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:name) do
+    isnamevar
+  end
+end
+```
+
+As mentioned earlier, if a type specifies a parameter named `:name`, it will automatically be the namevar, so the call to `isnamevar` is redundant, but is explicit.
+
+Alternatively, you can pass an option when calling newparam:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newparam(:name, :namevar => true) do
+  end
+end
+```
+
+Note that the namevar is necessarily a parameter, and not a property, since changing the name identifies a different resource as opposed to changing the name of an existing resource.
+
+Defining a parameter as the namevar also means it is required. Normally the namevar is automatically set to be the same as the title, unless the type overrides the `title_patterns` method, e.g. for composite namevars.
+
+### validate
+
+Validates an individual parameter value at catalog application time. The value to validate is yielded to the block:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:region) do
+    validate do |value|
+      fail 'region should not contain spaces' if value =~ /\s/
+    end
+  end
+end
+```
+
+Puppet parameter and properties can be multi-valued. For example, the `ec2_instance` can be given a list of `security_groups`, and each value in the list will be yielded to the `validate` method:
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  newproperty(:security_groups, ...) do
+    validate do |value|
+      fail 'security_groups should be ...' if value !~ /something/
+    end
+  end
+end
```

Comment:
yep that's mentioned above in https://github.com/puppetlabs/puppet-specifications/pull/72/files/1dd7369b7245c8f4b769a765976408583d426d70#diff-d666e1f4c3db61661d95e6cc0727a82dR102